resources/type-system/flow-analysis.md

```diff
@@ -414,39 +414,67 @@ Policy:
     - and `T <: S` or (`S` is `X extends R` and `T <: R`) or (`S` is `X & R` and
       `T <: R`)
 
-  - We say that a variable `x` is promotable via assignment of an expression of
-    type `T` given variable model `VM` if
-    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
-    - and `captured` is false
-    - and `S` is the current type of `x` in `VM`
-    - and `T <: S` and not `S <: T`
-    - and `T` is a type of interest for `x` in `tested`
-
-  - We say that a variable `x` is demotable via assignment of an expression of
-    type `T` given variable model `VM` if
-    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
-    - and `captured` is false
-    - and [...promoted, declared] contains a type `S` such that `T` is `S` or
-      `T` is **NonNull(`S`)**.
-
 Definitions:
 
+- `currentType(declared, promoted)`, where `declared` is a type and `promoted`
+  is a list of types, is the type `T` , defined as follows:
+  - If `promoted` is `[]`, `T` is `declared`.
+  - Otherwise, `T` is the last type in the `promoted` list.
+
+- `demote(promoted, written)`, where `promoted` is a list of types and `written`
+  is a type, is a list obtained by deleting any elements from `promoted` that do
+  not satisfy `written <: T`. _In effect, this removes any type promotions that
+  are no longer valid after the assignment of a value of type `written`._
+
+- `toi_promote(declared, demoted, tested, written)`, where `demoted` and
+  `tested` are lists of types, and `declared` and `written` are types, is the
+  list `promoted`, defined as follows. _("toi" stands for "type of interest".)_
+  - Let `current = currentType(declared, demoted)`. _(This is the type of the
+    variable after demotions, but before type of interest promotion.)_
+  - If `written` and `current` are the same type, then `promoted` is
+    `demoted`. _(No type of interest promotion is necessary in this case.)_
+  - Otherwise _(when `written` is not `current`)_, let `p1` be a set containing
+    the following types:
+    - **NonNull**(`declared`), if it is not the same as `declared`.
+    - For each type `T` in the `tested` list:
+      - `T`
+      - **NonNull**(`T`)
+
+    _The types in `p1` are known as the types of interest._
+  - Let `p2` be the set `p1 \ { current }` _(where `\` denotes set difference)_.
+  - If the `written` type is in `p2`, and `written <: current`, then `promoted`
+    is `[...demoted, written]`. _Writing a value whose static type is a type of
+    interest promotes to that type._
+    - _Since `written` is the type assigned to the variable (after type
+      coercion), in non-erroneous code it is guaranteed to be a subtype of
+      `declared`. And `toi_promote` is always applied to the result of `demote`,
```

Comment:
I think we should require at the beginning of the specification of `toi_promote` that `[declared, ...demoted]` is strictly decreasing, and `written` is a subtype of the last element in that list (not necessarily a proper subtype). We would then have that property here, and it could be checked for all invocations of `toi_promote` (of which we may have exactly one, but it generalizes easily if needed).

Reply:
> I think we should require at the beginning of the specification of `toi_promote` that `[declared, ...demoted]` is strictly decreasing, and `written` is a subtype of the last element in that list (not necessarily a proper subtype). We would then have that property here, and it could be checked for all invocations of `toi_promote` (of which we may have exactly one, but it generalizes easily if needed).

Done.